scripts/convertMarkdown.ts

```diff
@@ -0,0 +1,320 @@
+import {
+    convertDigitsInStringToDefaultNumeralSystem,
+    getIntFromNumberString
+} from './numeralUtils';
+import {
+    containsRomanScriptLetter,
+    getFilenameExt,
+    getFirstDigitsAsInt,
+    getIntFromString,
+    isBlank,
+    isNotBlank,
+    isPositiveInteger,
+    splitString,
+    stripAllExceptDigitsAndHyphens
+} from './stringUtils';
+
+export function convertMarkdownsToMilestones(
+    content: string,
+    docSet: string,
+    bookid: string
+): string {
+    let result: string = '';
+    result = content;
+    const sb = [];
+    let inputString = content;
+    const patternString = /(!?)\[([^[]*?)\]\((.*?)\)/;
+    let match;
+    while ((match = patternString.exec(inputString)) !== null) {
+        // Append text segment with 1st part of string
+        sb.push(inputString.substring(0, match.index));
+        // Handle markdown
+        const excl = match[1];
+        const text = match[2];
+        const ref = match[3];
+        const link = ref;
+        if (isBlank(ref)) {
+            // Empty link reference, e.g. [text]()
+            // Output simple text without a link
+            sb.push(text);
+        } else if (isLocalAudioFile(ref)) {
+            const zaudioc = getAudioHtmlFromMarkdownLink(link, text);
+            sb.push(zaudioc);
+        } else if (isImageLink(ref, excl)) {
+            // Image ![alt text](image.png)
+            const fig = getImageHtmlFromMarkdownLink(ref, text);
+            sb.push(fig);
+        } else if (isWebLink(ref)) {
+            const webLink = getWebHtmlFromMarkdownLink(link, text);
+            sb.push(webLink);
+        } else if (isEmailLink(ref)) {
+            const emailLink = getEmailHtmlFromMarkdownLink(link, text);
+            sb.push(emailLink);
+        } else if (isTelephoneNumberLink(ref)) {
+            const telLink = getTelHtmlFromMarkdownLink(link, text);
+            sb.push(telLink);
+        } else {
+            const refLink = getReferenceHtmlFromMarkdownLink(link, text, docSet, bookid);
+            sb.push(refLink);
+        }
+        inputString = inputString.substring(match.index + match[0].length);
+    }
+    sb.push(inputString);
+    result = sb.join('');
+    return result;
+}
+
+function isEmailLink(ref: string): boolean {
+    const refLower = ref.toLowerCase();
+    return refLower.startsWith('mailto:');
+}
+function isWebLink(ref: string): boolean {
+    const refLower = ref.toLowerCase();
+    return refLower.startsWith('http');
+}
+function isTelephoneNumberLink(ref: string): boolean {
+    const refLower = ref.toLowerCase();
+    return refLower.startsWith('tel:');
+}
+function isLocalAudioFile(ref: string): boolean {
+    let result = false;
+    const refLower = ref.toLowerCase();
+    if (!refLower.startsWith('http')) {
+        const ext = getFilenameExt(refLower);
+        if (ext != null) {
+            result = ext === 'mp3' || ext === 'ogg' || ext === 'wav';
```

Comment:
Does the Android code have .ogg?  I would have thought it would be .webm.

Reply:
Actually, for the web, Safari can't play .ogg or .webm files. So this should probably be just mp3 and wav.

Reply:
That is the way it is coded in BaseWriter.java.  I made the change to take that out of the pwa code.

Reply:
Changed it slightly to acknowledge webm and ogg links as audio links, but then to ignore them, not producing a link html since the file type is not supported